authorize/evaluator/policy_evaluator.go

```diff
@@ -0,0 +1,183 @@
+package evaluator
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/open-policy-agent/opa/rego"
+
+	"github.com/pomerium/pomerium/config"
+	"github.com/pomerium/pomerium/internal/log"
+	"github.com/pomerium/pomerium/pkg/policy"
+)
+
+// PolicyRequest is the input to policy evaluation.
+type PolicyRequest struct {
+	HTTP                     RequestHTTP    `json:"http"`
+	Session                  RequestSession `json:"session"`
+	IsValidClientCertificate bool           `json:"is_valid_client_certificate"`
+}
+
+// PolicyResponse is the result of evaluating a policy.
+type PolicyResponse struct {
+	Allow bool
+	Deny  *Denial
+}
+
+// Merge merges another PolicyResponse into this PolicyResponse. Access is allowed if either is allowed. Access is denied if
+// either is denied. (and denials take precedence)
+func (res *PolicyResponse) Merge(other *PolicyResponse) *PolicyResponse {
+	merged := &PolicyResponse{
+		Allow: res.Allow || other.Allow,
+		Deny:  res.Deny,
+	}
+	if other.Deny != nil {
+		merged.Deny = other.Deny
```

Comment:
should it be only if other.Deny is false?

Reply:
Deny is not a boolean. It includes an http status code and reason. A non-nil deny means the request should be denied.